### PR TITLE
Implement POUNDERs.

### PR DESCRIPTION
**Pull request by tobiasraabe**

### CODE REVIEW

File: estimagic/optimization/tao_optimizers.py [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+):
+    r"""Minimize a function using the POUNDERs algorithm.
+
+    POUNDERs can be a useful tool for economists who estimate structural models using
+    indirect inference, because unlike commonly used algorithms such as Nelder-Mead,
+    POUNDERs is tailored for minimizing a non-linear sum of squares objective function,
+    and therefore may require fewer iterations to arrive at a local optimum than
+    Nelder-Mead.
+
+    The criterion function :func:`func` should return an array of the errors NOT an
```
Comment (janosg) [inline estimagic/optimization/tao_optimizers.py:39]: This should probably explain the new dictionary interface

File: estimagic/optimization/tao_optimizers.py [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+    References:
+    .. _TAO Users Manual (Revision 3.7):
+        http://web.mit.edu/tao-petsc_v3.7/tao_manual.pdf
+    .. _Solving Derivative-Free Nonlinear Least Squares Problems with POUNDERS:
+        https://www.mcs.anl.gov/papers/P5120-0414.pdf
+    .. _petsc4py on BitBucket:
+        https://bitbucket.org/petsc/petsc4py
+
+    """
+    if sys.platform == "win32":
```
Comment (janosg) [inline estimagic/optimization/tao_optimizers.py:102]: We should replace this by a check that petsc4py is installed and have installation instructions in the error message.

File: estimagic/optimization/tao_optimizers.py [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+        raise NotImplementedError("The pounders algorithm is not available on Windows.")
+
+    func = functools.partial(
+        criterion_and_derivative, task="criterion", algorithm_info=POUNDERS_ALGO_INFO,
+    )
+
+    x = _initialise_petsc_array(x)
+    # We need to know the number of contributions of the criterion value to allocate the
+    # array.
+    n_errors = len(
```
Comment (janosg) [inline estimagic/optimization/tao_optimizers.py:112]: Very cool solution!

File: estimagic/optimization/tao_optimizers.py (hunk 2 above)
Comment (tobiasraabe) [reply estimagic/optimization/tao_optimizers.py:102]: Would you like to collect all `is_xxx_installed` checks in the configuration file? The variable would be necessary in the optimizer packages and the tests and thus we could have a central place for it.

Comment (janosg) [reply estimagic/optimization/tao_optimizers.py:102]: Yes!

**Comment by tobiasraabe** [comment]
@janosg seems fine.

**Review by janosg — APPROVED** [review]
Yes, you can merge and make a dev release.

**Pull request merged**